### PR TITLE
Update search.cc to implement more aggressive time prune threshold to CLOP tune

### PR DESCRIPTION
**Pull request by jjoshua2**
Implement EARLY-C strategy from https://pdfs.semanticscholar.org/a2e6/299fd3c8ab17e3a1a783d518688b55bb2363.pdf. I came up with this independently and tried it in lczero as well https://github.com/glinscott/leela-chess/pull/556/files and found it helped CPU but not GPU for some reason...

### CODE REVIEW

File: src/mcts/search.cc [hunk 1]
```diff
@@ -41,6 +41,7 @@ const char* Search::kTempDecayMovesStr = "Moves with temperature decay";
 const char* Search::kNoiseStr = "Add Dirichlet noise at root node";
 const char* Search::kVerboseStatsStr = "Display verbose move stats";
 const char* Search::kSmartPruningStr = "Enable smart pruning";
+const char* Search::pEarlyExit = "Aggressive smart pruning threshold";
```
Comment (mooskagh) [inline src/mcts/search.cc]: pEarlyExit**Str**

File: src/mcts/search.cc [hunk 2]
```diff
@@ -72,6 +73,8 @@ void Search::PopulateUciParams(OptionsParser* options) {
   options->Add<BoolOption>(kNoiseStr, "noise", 'n') = false;
   options->Add<BoolOption>(kVerboseStatsStr, "verbose-move-stats") = false;
   options->Add<BoolOption>(kSmartPruningStr, "smart-pruning") = true;
+  options->Add<FloatOption>(p_early_exit, 0.1f, 10.0f,
+                            "p_early_exit") = 1.0f;
```
Comment (mooskagh) [inline src/mcts/search.cc]: Dashes in flag name instead of underscores.

File: src/mcts/search.cc [hunk 3]
```diff
@@ -578,7 +581,7 @@ SearchWorker::NodeToProcess SearchWorker::PickNodeToExtend() {
         // To ensure we have at least one node to expand, always include
         // current best node.
         if (child != search_->best_move_node_ &&
-            search_->remaining_playouts_ <
+            p_early_exit*search_->remaining_playouts_ <
```
Comment (mooskagh) [inline src/mcts/search.cc]: Where is this variable defined?
Also it's one additional float multiplication per visit. Probably not noticeable at all, but would be interesting to  check with `--backend=random`

File: src/mcts/search.cc [hunk 4]
```diff
@@ -72,6 +73,8 @@ void Search::PopulateUciParams(OptionsParser* options) {
   options->Add<BoolOption>(kNoiseStr, "noise", 'n') = false;
   options->Add<BoolOption>(kVerboseStatsStr, "verbose-move-stats") = false;
   options->Add<BoolOption>(kSmartPruningStr, "smart-pruning") = true;
+  options->Add<FloatOption>(p_early_exit, 0.1f, 10.0f,
```
Comment (mooskagh) [inline src/mcts/search.cc]: pEarlyExitStr

**Comment by jjoshua2** [comment]
The names of the constant p_early_exit is taken from the paper. It's the probability or really percentage that you prune at. If you prune at 1.0 (or above) its 100% chance of success not missing anything given a max_time, whereas the paper actually find optimal was p = 0.4 where you prune at only 40% of the nodes required to change it's mind, and increase scale_mover from 2 to 2.5.

**Comment by jjoshua2** [comment]
I'm still trying different values but it seems like a 20 elo gain is possible given my two results ```Score of lc0_PR123 tuned vs lc0_PR123 d: 61 - 50 - 142 [0.522]
Elo difference: 15.12 +/- 28.37``` and ```18s
Score of lc0_PR123 tuned vs lc0_PR123 default: 129 - 102 - 418 [0.521]
Elo difference: 14.46 +/- 15.92```

**Comment by jjoshua2** [comment]
0.72 kEarly and 2.5 scale time with latest at the time main net 520
```
Rank Name                          Elo     +/-   Games   Score   Draws
   0 Ethereal10.66 16CPU 4GB TB       8      46     130   51.2%   40.8%
   1 lc0_PR123 520 tuned             5      62      66   50.8%   47.0%
   2 lc0_PR123 520                 -22      70      64   46.9%   34.4%
```

**Comment by jjoshua2** [comment]
I put in the average for the max and mean from the clop tune this morning. I'm not planning on CLOP tuning any more until after TCEC starts probably as I'm testing latest testnets to send now.

**Comment by jjoshua2** [comment]
Another test using roy's PR testing .68 vs 1.0 pruning
```Score of lc0 10810 leroy j vs lc0 10810 1.0 asp: 49 - 31 - 402 [0.519]
Elo difference: 12.98 +/- 12.59```

**Pull request merged**